readme.txt
JAdvisor - a Java class scheduler and course planner for students.
jadvisor.sourceforge.net

LICENSE

Copyright (C) 2001-2002 Curtis Rawls

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

USAGE

To run:

	Windows:
	Click on jadvisor.jar file (in /bin)
		or
	Run jadvisor.bat
	
	Linux:
	Run jadvisor.sh
	
	Command Line:
	cd bin
	java jadvisor/advisorui/Advisor [filename]

	From the JAR file:
	java -jar jadvisor.jar [filename]

To compile:
	cd src
	javac -d ../bin jadvisor/advisorui/Advisor.java

To make a JAR file:
	cd bin
	jar cmvf mainClass jadvisor.jar jadvisor icons

SCHOOL ADAPTERS

JAdvisor uses SchoolAdapters to customize itself for a student's
school.  Currently NC State University is the only implemented
SchoolAdapter for a specific college.  To customize JAdvisor for
your own school, just implement the provided SchoolAdapter interface.
For help writing your own SchoolAdapter, check out the NCSU adapter 
for example code and post to the JAdvisor forums with your questions.
Once your SchoolAdapter is done, let us know so other students at your
school can take advantage of JAdvisor!

changelog.txt
JAdvisor Changelog
0.4.6
 SchoolAdapter throws new Exceptions
 Added SplitPanes to SchedulerUI and PlannerUI
 PlannerUI displays semester names
 Text Box bug fixed
 Scheduler UI bugfixes
 NCSU adapter bugfixes
 UNC adapter updated
0.4.5
 Improved course selection in Planner (PlanViewPanel changes)
 Created AdvisorDialogs class
 Added threading to SchedulerUI
 Added threading to SearchUI
0.4.4
 Scheduler displays Blocks in the table
 Fix for ScheduleGraph bug
 Improved StudentBlockUI
0.4.3
 Added ability to save your schedule or plan to HTML
 Moved properties file to "user.home"/.jadvisor
0.4.2
 Added New Student Wizard
 Added Preferences menu
 Curriculum list in Planner displays courses
 Tooltips for prerequisites in Planner
 Fixed Save as HTML in ScheduleGraph
 SchoolAdapter for University of North Carolina now functional
 Bugfixes to NCSU Adapter
0.4.1
 New close dialogs
 New SchoolAdapter choice dialog
 Bug fixes
0.4.0
 New .exe installer for Windows users
 ScheduleGraphUI outputs graph as HTML
 Increased customization with SchoolAdapters
 Better string output of TimeOfDay
0.3.1
 ScheduleGraph now displays an HTML graph of courses
 SchoolAdapter for NC State University now functional
 Better interface specs for SchoolAdapter formalized
0.3.0
 Bug fixes for Schedule Wizard
 Added toString for StudentSemesterPlan and StudentSchedule
 Added splash window
 Improved support for icons
 Bug fixes in ScheduleGraphUI
 Saves window properties
 About dialog
0.2.0
 Full file support - new/open/save .jaf files
0.1.1
 Added JA icon
 Better new/open/save file support (still not finished)
 File options refresh the GUI
 Created Advisor.java, moved main to here
 Start properties support
 Title displays opened file's name
0.1.0
 Initial release

src/jadvisor/advisorui/AdvisorUI.java
/*
 * jadvisor/advisorui/AdvisorUI.java - JAdvisor UI
 * Copyright (C) 2001-2002 Curtis Rawls
 * jadvisor.sourceforge.net
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */

package jadvisor.advisorui;

import jadvisor.planner.*;
import jadvisor.scheduler.*;
import jadvisor.school.*;

import javax.swing.*;
import javax.swing.event.*;
import javax.swing.filechooser.*;
import java.awt.*;
import java.awt.event.*;
import java.util.*;
import java.io.File;
import java.net.URL;

/**
 * AdvisorUI extends the <code>JFrame</code> class.  It manages all the UI 
 * elements of JAdvisor: menu, toolbar, scheduler, planner, search, and message
 * field.
 *
 * @author Curtis Rawls
 * @version 0.4.6
 */
public class AdvisorUI extends JFrame {
	private static final String ICONFILENAME = "/icons/ja.gif";
	private static final int SPLASHWAIT = 5000;//5000 == 5 seconds
	
	private final Advisor _advisor;
	private final AdvisorMenu _menu;
	private final JToolBar _toolBar;
	private SchedulerUI _scheduler;
	private PlannerUI _planner;
	private SearchUI _search;
	private final JTextArea _textArea;
	private final Component _textAreaComponent;
	private SchoolAdapter _schoolAdapter;
	private Properties _properties;

public AdvisorUI(Advisor advisor, Properties properties) {
	super("JAdvisor");
    
    _advisor = advisor;
	_properties = properties;

	SplashWindow splash;
	if (_properties.getProperty("showsplash", "true").equals("true"))
		splash = new SplashWindow(this, SPLASHWAIT);//draws splash window

	//sets Look and Feel
	try {
        UIManager.setLookAndFeel(
	    	UIManager.getSystemLookAndFeelClassName());
	    //UIManager.getCrossPlatformLookAndFeelClassName());
    } catch (Exception e) { }
    
    load();

	//OptionsUI.initialize(this, this, _schoolAdapter, (Properties)_properties.clone());
	
	_menu = new AdvisorMenu(advisor, this);
	_toolBar = toolBar();
	_textArea = textArea();
	_textAreaComponent = new JScrollPane(_textArea,
			JScrollPane.VERTICAL_SCROLLBAR_ALWAYS, 
			JScrollPane.HORIZONTAL_SCROLLBAR_AS_NEEDED);
	
	frame();
}

private void load () {
}

private void frame () {
	setDefaultCloseOperation(JFrame.DO_NOTHING_ON_CLOSE);
	
	setJMenuBar(_menu);
	getContentPane().setLayout(new BorderLayout(0, 3));
	
	try {
		//setIconImage(Toolkit.getDefaultToolkit().getImage(ICONFILENAME));
		setIconImage(Toolkit.getDefaultToolkit().getImage(getClass().getResource(ICONFILENAME)));
		//setIconImage(FileUtility.getImageFromJAR(ICONFILENAME, getClass()));
	} catch (NullPointerException e) {System.err.println("Could not open " + ICONFILENAME + ".");}
	
	addWindowListener(exitAction());
	setSize(Integer.parseInt(_properties.getProperty("width", "600")), 
		Integer.parseInt(_properties.getProperty("height", "450")));
	setLocation(Integer.parseInt(_properties.getProperty("xloc", "0")), 
		Integer.parseInt(_properties.getProperty("yloc", "0")));
}

private WindowListener exitAction () {
	return new WindowAdapter () {
		public void windowClosing(WindowEvent e) {
			_advisor.exit();}
	};
}

private JToolBar toolBar () {
	final JToolBar result = new JToolBar();
	final JButton scheduleButton = new JButton("Schedule");
	final JButton scheduleGraphButton = new JButton("Schedule Graph");
	scheduleGraphButton.addActionListener(new ActionListener() {
		public void actionPerformed(ActionEvent e) {
			ScheduleGraphUI.showDialog();
		}
	});
	final JButton planButton = new JButton("Plan");
	planButton.addActionListener(new ActionListener() {
		public void actionPerformed(ActionEvent e) {
		}
	});
	//result.add(scheduleButton);
	result.add(scheduleGraphButton);
	//result.add(planButton);
	result.setFloatable(false);
	return result;
}

private JTextArea textArea () {
	JTextArea result = new JTextArea(3, 30);
	result.setEditable(false);
	return result;
}

void open (StudentFile studentFile, File file, SchoolAdapter schoolAdapter) {
	_schoolAdapter = schoolAdapter;
	_scheduler = new SchedulerUI(studentFile.getStudentSchedule(),
		studentFile.getStudentPlan(), studentFile.getStudentInfo(), 
		schoolAdapter, _properties);

	_planner = new PlannerUI(this, studentFile.getStudentPlan(), studentFile.getStudentInfo(), schoolAdapter);
	_search = new SearchUI(schoolAdapter);
	final JTabbedPane tabbedPane = new JTabbedPane();
	tabbedPane.addTab("Scheduler", _scheduler);
	tabbedPane.addTab("Planner", _planner);
	tabbedPane.addTab("Search", _search);
	tabbedPane.setSelectedIndex(0);
	getContentPane().removeAll();
	getContentPane().add(_toolBar, BorderLayout.NORTH);
	getContentPane().add(tabbedPane, BorderLayout.CENTER);
	if (_properties.getProperty("showtextbox", "true").equals("true"))
		getContentPane().add(_textAreaComponent, BorderLayout.SOUTH);
	show();
	setTitle(file);
	
	ScheduleGraphUI.initialize(studentFile.getStudentSchedule(), _properties, schoolAdapter);

	monitorStudentSchedule(studentFile.getStudentSchedule());
	monitorStudentPlan(studentFile.getStudentPlan());
}

void setTitle (File file) {
	if (file == null)
		setTitle("JAdvisor - [untitled]");
	else
		setTitle("JAdvisor - [" + file + "]");
}

void setTextAreaVisible (boolean visible) {
	if (visible)
		getContentPane().add(_textAreaComponent, BorderLayout.SOUTH);
	else
		getContentPane().remove(_textAreaComponent);
	show();
}

void saveProperties (Properties p) {
	p.setProperty("xloc", "" + getX());
	p.setProperty("yloc", "" + getY());
	p.setProperty("width", "" + getWidth());
	p.setProperty("height", "" + getHeight());
	_scheduler.saveProperties(p);
	ScheduleGraphUI.saveProperties(p);
}

void showAboutDialog () {
	JPanel aboutPanel = new JPanel();
	aboutPanel.setLayout(new BoxLayout(aboutPanel, BoxLayout.Y_AXIS));
	aboutPanel.add(new JLabel("JAdvisor - a Java class scheduler and course planner for students."));
	aboutPanel.add(new JLabel("Copyright 2001-2002 Curtis Rawls"));
	aboutPanel.add(new JLabel("All rights reserved."));
	aboutPanel.add(new JLabel("jadvisor.sourceforge.net"));
	JOptionPane optionPane = new JOptionPane(aboutPanel, 
		JOptionPane.INFORMATION_MESSAGE, JOptionPane.DEFAULT_OPTION, 
		new ImageIcon(Toolkit.getDefaultToolkit().getImage(getClass().getResource(ICONFILENAME))));
	JDialog dialog = optionPane.createDialog(this, "About JAdvisor v0.4.6");
	dialog.show();
}

private void monitorStudentSchedule(StudentSchedule p) {
	p.addListener(new StudentScheduleListener() {
		public void classAdded (StudentSchedule s, StudentClass c) {
			_textArea.append("Class added: " + _schoolAdapter.classToString(c) + "\n");
		}
		public void classRemoved (StudentSchedule s, StudentClass c) {
			_textArea.append("Class removed: " + _schoolAdapter.classToString(c) + "\n");
		}
		public void blockAdded (StudentSchedule s, StudentBlock b) {
			_textArea.append("Block added: " + b.toString() + "\n");
		}
		public void blockRemoved (StudentSchedule s, StudentBlock b) {
			_textArea.append("Block removed: " + b.toString() + "\n");
		}
	});
}

private void monitorStudentPlan(StudentPlan p) {
	p.addListener(new StudentPlanListener() {
		public void courseAdded (StudentPlan s, Course c, int year, int semester) {
			_textArea.append("Course added: " + c.toString() + " to " 
				+ year + " " + _schoolAdapter.getSemesters()[semester] + "\n");
		}
		public void courseRemoved (StudentPlan s, Course c, int year, int semester) {
			_textArea.append("Course removed: " + c.toString() + " from " 
				+ year + " " + _schoolAdapter.getSemesters()[semester] + "\n");
		}
	});
}

}
